### PR TITLE
Reuse the proxy's build stage across CI runs

### PR DESCRIPTION
**Pull request by olix0r**
The proxy's Dockerfile is split into stages: build and runtime.
The build stage includes all of the intermdiate build information, and
the runtime image discards these layers with a small production-ready
image.

In order to improve docker build times, we can save this build layer to
be reused.

This reduces the docker build of the proxy in CI from 15 minutes to
about 7.5 minutes (when the proxy is not changed).

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -124,11 +125,13 @@ jobs:
...
         - export BUILD_DEBUG=1 DOCKER_TRACE=1
+        - export DOCKER_CACHE="$HOME/.cache/docker.tgz" ; if [ -f "$DOCKER_CACHE" ]; then gunzip -c "$DOCKER_CACHE" | docker load ; fi
+        - docker image ls gcr.io/runconduit/proxy-build:cache
 
       script:
-        # We re-run the tests here in *release* (--release) mode since the
-        # `test` stage only runs them in *debug* mode.
         - bin/docker-build
+        # The proxy's build cache is preserved so it can be reused on subsequent runs.
+        - docker save gcr.io/runconduit/proxy-build:cache | gzip >"$DOCKER_CACHE"
```
Comment (briansmith) [inline .travis.yml]: TIOLI: `gzip -9`

File: .travis.yml [hunk 2]
```diff
@@ -124,11 +125,13 @@ jobs:
           export CONDUIT_TAG=$(. bin/_tag.sh ; clean_head_root_tag)
           echo "CONDUIT_TAG=${CONDUIT_TAG}"
         - export BUILD_DEBUG=1 DOCKER_TRACE=1
+        - export DOCKER_CACHE="$HOME/.cache/docker.tgz" ; if [ -f "$DOCKER_CACHE" ]; then gunzip -c "$DOCKER_CACHE" | docker load ; fi
```
Comment (briansmith) [inline .travis.yml]: Separate the `export DOCKER_CACHE` to its own statement like the line above.

File: bin/docker-build-proxy [hunk 3]
```diff
@@ -13,5 +13,14 @@ rootdir="$( cd $bindir/.. && pwd )"
 . $bindir/_docker.sh
 . $bindir/_tag.sh
 
+# The proxy's build stage may be useful to preserve, especially in CI. We tag this stage
+# separately. The subsequent runtime build is trivial.
+docker_build proxy-build cache $rootdir/proxy/Dockerfile \
+    --build-arg="PROXY_UNOPTIMIZED=${PROXY_UNOPTIMIZED:-}" \
+    --cache-from="$(docker_repo proxy-build):cache" \
```
Comment (briansmith) [inline bin/docker-build-proxy]: I'm not convinced this is the right behavior outside of Travis CI runs due to https://github.com/moby/moby/issues/34715. My understanding is that if we specify `--cache-from` here then that will disable any/all implicit local caching. IIUC, without `--cache-from`, I will use many different cached versions of the build container locally, as needed for the different branches I have, but with `--cache-from` I will only be able to used one cached version. In particular, if I am switching between different branches with `git checkout` on my local machine that require a different build image than the one specified in `--cache-from` above, then no caching will be done. So, unless I'm misunderstanding, I think we should only do this `--cache-from` in CI and not by default.

Comment (olix0r) [reply bin/docker-build-proxy]: yeah, this makes sense.

File: bin/docker-build-proxy [hunk 4]
```diff
@@ -13,5 +13,15 @@ rootdir="$( cd $bindir/.. && pwd )"
...
+    # If a cache is not specified use the default docker cache.
+    docker_build proxy "$(head_root_tag)" $rootdir/proxy/Dockerfile \
+        --build-arg="PROXY_UNOPTIMIZED=${PROXY_UNOPTIMIZED:-}"
+
+else
+    # Otherwise, if a cache image is specified, use it.
+    docker_build proxy "$(head_root_tag)" $rootdir/proxy/Dockerfile \
+        --build-arg="PROXY_UNOPTIMIZED=${PROXY_UNOPTIMIZED:-}" \
+        --cache-from="$PROXY_BUILD_CACHE_IMAGE" \
+        --target runtime
```
Comment (briansmith) [inline bin/docker-build-proxy]: NIT: Isn't `--target runtime` the default? It is a bit confusing that we use `--target runtime` in the `else` branch but not the `if` branch.

**Pull request merged**